File: 977-squares-of-a-sorted-array/977-squares-of-a-sorted-array.cpp
#include <stdlib.h>     /* abs */
#include <vector>

class Solution {
public:
    vector<int> sortedSquares(vector<int>& nums) {
        
        /**
        * This naive approach is O(nlogn) due to sorting
        **/
//         vector<int> squared;
//         for (auto& num : nums) {
//             squared.push_back(num*num);
//         }
        
//         std::sort(squared.begin(), squared.end());
        
//         return squared;
        
        /**
        * This approach is O(n)
        **/
        
        vector<int> squared;
        int l = 0;
        int r = nums.size() - 1;
        
        if (nums.size() == 0) return squared;
        
        while (l < r) {
            if (abs(nums[l]) > abs(nums[r])) {
                squared.insert(squared.begin(), nums[l] * nums[l]); // vector has no 'push_front' 
                l++;
            }
            
            else if (abs(nums[l]) <= abs(nums[r])) {
                squared.insert(squared.begin(), nums[r] * nums[r]);
                r--;
            }
            
        }

        squared.insert(squared.begin(), nums[l] * nums[l]);
        
        return squared;
    }
};

File: 977-squares-of-a-sorted-array/977-squares-of-a-sorted-array.py
class Solution:
    def sortedSquares(self, nums: List[int]) -> List[int]:
        l,r = 0,len(nums)-1
        
        res = deque()
        
        while l <= r:
            if abs(nums[l]) < abs(nums[r]):
                res.appendleft(nums[r] ** 2)
                r-=1
            else:
                res.appendleft(nums[l] ** 2)
                l+=1
        return res